File: wavexlr/icons.py
"""Resolve icons at draw time without changing persisted source/mix choices."""

from functools import lru_cache
import logging
import os

from . import paths

TRAY_ICON_COLORS = ("white", "black")

# Import GTK only when a display is available; icon metadata is also used headless.
_ALTERNATIVES = {
    "web-browser-symbolic": (
        "internet-web-browser-symbolic",
        "applications-internet-symbolic",
        "globe-symbolic",
    ),
    "input-gaming-symbolic": (
        "applications-games-symbolic",
        "input-gamepad-symbolic",
    ),
    "audio-x-generic-symbolic": (
        "multimedia-player-symbolic",
        "media-optical-audio-symbolic",
    ),
    "list-drag-handle-symbolic": (
        "view-list-symbolic",
        "open-menu-symbolic",
    ),
    "network-transmit-symbolic": (
        "network-wired-symbolic",
        "network-connect-symbolic",
    ),
    "preferences-desktop-multimedia-symbolic": (
        "multimedia-player-symbolic",
        "applications-multimedia-symbolic",
    ),
    "video-display-symbolic": (
        "computer-symbolic",
        "preferences-desktop-display-symbolic",
    ),
}

_cache = {}
_watched = False


def _theme():
    """The display's icon theme, or None when there is no display yet."""
    global _watched
    try:
        import gi
        gi.require_version("Gtk", "4.0")
        from gi.repository import Gdk, Gtk
    except (ImportError, ValueError):
        return None
    display = Gdk.Display.get_default()
    if display is None:
        return None
    theme = Gtk.IconTheme.get_for_display(display)
    if theme is not None and not _watched:
        # A theme change makes every earlier answer stale, including the ones
        # that needed no substitution.
        theme.connect("changed", lambda *_: _cache.clear())
        _watched = True
    return theme


def resolve(name):
    """Return name, or the nearest name the active theme actually has.

    Unknown names are returned untouched: a theme we have no table for is not
    improved by guessing, and the broken glyph is at least honest about it.
    """
    if not name:
        return name
    if name in _cache:
        return _cache[name]

    theme = _theme()
    if theme is None:
        return name
    chosen = name
    if theme is not None and not theme.has_icon(name):
        for alternative in _ALTERNATIVES.get(name, ()):
            if theme.has_icon(alternative):
                chosen = alternative
                break

    _cache[name] = chosen
    return chosen


@lru_cache(maxsize=None)
def _asset_path(name):
    path = paths.data_file("icons", f"{name}.svg")
    if path is None:
        logging.warning("icons: supplied artwork %s.svg is missing", name)
    return path


def theme_path(name):
    """Directory containing supplied artwork, or empty when it is missing."""
    path = _asset_path(name)
    return os.path.dirname(path) if path else ""

File: crates/openwave-desktop/src/icons.rs
use openwave_core::model::{OperationError, Result};
use openwave_runtime::paths::RuntimePaths;
use std::{
    cell::{Cell, RefCell},
    collections::HashMap,
    path::PathBuf,
    rc::Rc,
};

/// Rendering choices never replace the icon name stored in a source or preference.
#[derive(Clone)]
pub struct Icons {
    supplied: Rc<HashMap<String, Result<PathBuf>>>,
    resolved: Rc<RefCell<HashMap<String, String>>>,
    watched: Rc<Cell<bool>>,
}

impl Icons {
    pub fn new(paths: RuntimePaths) -> Self {
        let supplied = [
            "openwave",
            "openwave-white",
            "openwave-black",
            "openwave-red",
        ]
        .into_iter()
        .map(|name| {
            (
                name.to_owned(),
                paths.data_file(&format!("icons/{name}.svg")),
            )
        })
        .collect();
        let resolved = Rc::new(RefCell::new(HashMap::new()));
        Self {
            supplied: Rc::new(supplied),
            resolved,
            watched: Rc::new(Cell::new(false)),
        }
    }

    pub fn supplied_path(&self, name: &str) -> Result<PathBuf> {
        self.supplied
            .get(name.strip_suffix(".svg").unwrap_or(name))
            .cloned()
            .unwrap_or_else(|| Err(OperationError::invalid("Not supplied OpenWave artwork")))
    }

    pub fn image(&self, name: &str, size: i32) -> gtk::Image {
        let image = if let Ok(path) = self.supplied_path(name) {
            gtk::Image::from_file(path)
        } else {
            gtk::Image::from_icon_name(&self.resolve(name))
        };
        image.set_pixel_size(size);
        image
    }

    pub(crate) fn theme_path(&self, name: &str) -> String {
        self.supplied_path(name)
            .ok()
            .and_then(|path| path.parent().map(|p| p.to_string_lossy().into_owned()))
            .unwrap_or_default()
    }

    pub(crate) fn resolve(&self, name: &str) -> String {
        if let Some(chosen) = self.resolved.borrow().get(name) {
            return chosen.clone();
        }
        let Some(display) = gtk::gdk::Display::default() else {
            return name.to_owned();
        };
        let theme = gtk::IconTheme::for_display(&display);
        if !self.watched.replace(true) {
            let cache = Rc::downgrade(&self.resolved);
            theme.connect_changed(move |_| {
                if let Some(cache) = cache.upgrade() {
                    cache.borrow_mut().clear();
                }
            });
        }
        let alternatives: &[&str] = match name {
            "web-browser-symbolic" => &[
                "internet-web-browser-symbolic",
                "applications-internet-symbolic",
                "globe-symbolic",
            ],
            "input-gaming-symbolic" => &["applications-games-symbolic", "input-gamepad-symbolic"],
            "audio-x-generic-symbolic" => {
                &["multimedia-player-symbolic", "media-optical-audio-symbolic"]
            }
            "list-drag-handle-symbolic" => &["view-list-symbolic", "open-menu-symbolic"],
            "network-transmit-symbolic" => &["network-wired-symbolic", "network-connect-symbolic"],
            "preferences-desktop-multimedia-symbolic" => &[
                "multimedia-player-symbolic",
                "applications-multimedia-symbolic",
            ],
            "video-display-symbolic" => {
                &["computer-symbolic", "preferences-desktop-display-symbolic"]
            }
            _ => &[],
        };
        let chosen = if theme.has_icon(name) {
            name
        } else {
            alternatives
                .iter()
                .copied()
                .find(|candidate| theme.has_icon(candidate))
                .unwrap_or(name)
        }
        .to_owned();
        self.resolved
            .borrow_mut()
            .insert(name.to_owned(), chosen.clone());
        chosen
    }
}
